code/modules/power/solar.dm

```diff
@@ -318,95 +316,62 @@ TYPEINFO(/obj/machinery/power/solar)
 	gen = 0
 	SEND_SIGNAL(src,COMSIG_MECHCOMP_TRANSMIT_SIGNAL, "power=[num2text(round(lastgen), 50)]&powerfmt=[engineering_notation(lastgen)]W&angle=[cdir]")
 
-	if(status & (NOPOWER | BROKEN))
-		return
-
-	if(track==1 && nexttime < world.timeofday && trackrate)
-		nexttime = world.timeofday + 3600/abs(trackrate)
-		cdir = (cdir+trackrate/abs(trackrate)+360)%360
-
-		set_panels(cdir)
-
-	src.updateDialog()
-
 
 // called by solar tracker when sun position changes
 /obj/machinery/computer/solar_control/proc/tracker_update(var/angle)
-	if(track != 2 || status & (NOPOWER | BROKEN))
+	if(!active || status & (NOPOWER | BROKEN))
 		return
-	cdir = angle
+	cdir = emagged ? (angle + 180) % 360 : angle
 	set_panels(cdir)
 
-	src.updateDialog()
-
 /obj/machinery/computer/solar_control/attack_hand(mob/user)
 	if(..())
 		return
 
 	if ( (BOUNDS_DIST(src, user) > 0 ))
 		if (!isAI(user))
-			src.remove_dialog(user)
-			user.Browse(null, "window=solcon")
 			return
 
-	add_fingerprint(user)
-	src.add_dialog(user)
+	if (istype(user.equipped(), /obj/item/card/emag))
+		return
 
-	var/t = "<TT><B>XIANG|GIESEL Photo-Electric Generator Control</B><HR><PRE>"
-	t += "Generated power : [round(lastgen)] W<BR><BR>"
-	t += "<B>Orientation</B>: [rate_control(src,"cdir","[cdir]&deg",1,15)] ([angle2text(cdir)])<BR><BR><BR>"
+	active = !active
+	user.show_text("You [active ? "activate" : "deactivate"] [src]'s solar tracking.", "blue")
+	if (src.is_active_and_powered())
+		src.tracker_update(tracker.sun_angle)
 
-	t += "<BR><HR><BR><BR>"
+	src.update_solar_icon()
 
-	t += "Tracking: "
-	switch(track)
-		if(0)
-			t += "<B>Off</B> <A href='?src=\ref[src];track=1'>Timed</A> <A href='?src=\ref[src];track=2'>Auto</A><BR>"
-		if(1)
-			t += "<A href='?src=\ref[src];track=0'>Off</A> <B>Timed</B> <A href='?src=\ref[src];track=2'>Auto</A><BR>"
-		if(2)
-			t += "<A href='?src=\ref[src];track=0'>Off</A> <A href='?src=\ref[src];track=1'>Timed</A> <B>Auto</B><BR>"
+/obj/machinery/computer/solar_control/get_desc()
+	. = "<br />It is currently <em>[src.is_active_and_powered() ? "tracking the sun" : "disabled"]</em>"
+	. += "<br />Generated power: [round(lastgen)] W"
+	. += "<br />Current Orientation: [cdir]&deg; ([angle2text(cdir)])"
+	. += "<br />Sun Orientation: [tracker.sun_angle]&deg; ([angle2text(tracker.sun_angle)])"
 
+/obj/machinery/computer/solar_control/proc/is_active_and_powered()
+	. = active && !(status & (NOPOWER | BROKEN))
 
-	t += "Tracking Rate: [rate_control(src,"tdir","[trackrate] deg/h ([trackrate<0 ? "CCW" : "CW"])",5,30,180)]<BR><BR>"
-	t += "<A href='?src=\ref[src];close=1'>Close</A></TT>"
-	user.Browse(t, "window=solcon")
-	onclose(user, "solcon")
-	return
+/obj/machinery/computer/solar_control/proc/update_solar_icon()
+	if (src.is_active_and_powered())
+		src.icon_state = initial(src.icon_state)
+	else
+		src.icon_state = "solar0"
+	src.UpdateIcon()
```

Comment:
```suggestion
/obj/machinery/computer/solar_control/update_icon()
	. = ..()
	if (src.is_active_and_powered())
		src.icon_state = initial(src.icon_state)
	else
		src.icon_state = "solar0"
```
We can just roll this into an `update_icon` override instead of needing a new proc